File: 2018-05-20/2.cpp
#include <algorithm>
#include <cassert>
#include <cmath>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#define MAXN 210
#define EPS 1e-5
const double PI = acos(-1);
int n, m, k, r, _x[MAXN], _y[MAXN], x_[MAXN], y_[MAXN];
double ans, rd;
inline double S(double x) { return x * x * PI; }
inline double sqr(int x) { return 1. * x * x; }
inline double dist(int x0, int y0, int x1, int y1) { return sqrt(sqr(x0 - x1) + sqr(y0 - y1)); }
std::pair<double, std::pair<int, int>> dt[MAXN * MAXN];
int match[MAXN], vis[MAXN], skip[MAXN], ban[MAXN][MAXN];
bool _(int x) {
    for (int i = 1; i <= m; i++) {
        if (ban[x][i]) continue;
        if (vis[i]) continue;
        vis[i] = 1;
        if (!match[i] || _(match[i])) return match[i] = x, 1;
    }
    return 0;
}
int main() {
    scanf("%d%d%d", &n, &m, &r);
    for (int i = 1; i <= n; i++) scanf("%d", &_x[i]);
    for (int i = 1; i <= n; i++) scanf("%d", &_y[i]);
    for (int i = 1; i <= m; i++) scanf("%d", &x_[i]);
    for (int i = 1; i <= m; i++) scanf("%d", &y_[i]);
    for (int i = 1; i <= n; i++) {
        for (int j = 1; j <= m; j++) {
            double d = dist(_x[i], _y[i], x_[j], y_[j]) / 2;
            if (d < r + EPS) dt[++k] = {d, {i, j}};
        }
    }
    dt[++k] = {r, {0, 0}};
    std::sort(dt + 1, dt + k + 1);
    memset(ban, 1, sizeof(ban));
    int cnt = 0;
    for (int i = 1; i <= k; i++) {
        memset(skip, 0, sizeof(skip));
        for (int i = 1; i <= m; i++) skip[match[i]] = 1;
        for (int i = 1; i <= n; i++)
            if (!skip[i] && (memset(vis, 0, sizeof(vis)), _(i))) cnt++;
        int res = n + m - cnt;
        // printf("Radius %lf: count %d, ans %lf\n", dt[i].first, res, S(dt[i].first) * res);
        ans = std::max(ans, S(dt[i].first) * res);
        ban[dt[i].second.first][dt[i].second.second] = 0;
    }
    printf("%.4lf\n", ans);
    return 0;
}

File: 2018-05-20/2.java
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.ArrayList;

class data implements Comparable<data> {
    public double r;
    public int x, y;

    public data() {
    }

    public data(double r, int x, int y) {
        this.r = r;
        this.x = x;
        this.y = y;
    }

    @Override
    public int compareTo(data obj) {
        return this.r < obj.r ? 1 : 0;
    }
}

class HouseProtection {
    final int MAXN = 210;
    final double EPS = 1e-5;
    final double PI = Math.acos(-1);
    int n, m, k, r;
    int[] _x, _y, x_, y_;
    double ans, rd;

    double S(double x) {
        return x * x * PI;
    }

    double sqr(int x) {
        return 1. * x * x;
    }

    double dist(int x0, int y0, int x1, int y1) {
        return Math.sqrt(sqr(x0 - x1) + sqr(y0 - y1));
    }

    int match[] = new int[MAXN];
    boolean vis[] = new boolean[MAXN];
    boolean skip[] = new boolean[MAXN];
    boolean ban[][] = new boolean[MAXN][MAXN];

    boolean h(int x) {
        for (int i = 1; i <= m; i++) {
            if (!ban[x][i])
                continue;
            if (vis[i])
                continue;
            vis[i] = true;
            if (match[i] == 0 || h(match[i])) {
                match[i] = x;
                return false;
            }
        }
        return true;
    }

    public double safetyFactor(int _x[], int _y[], int x_[], int y_[], int R) {
        n = _x.length;
        m = x_.length;

        List<data> list = new ArrayList<data>();
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                double d = dist(_x[i], _y[i], x_[j], y_[j]) / 2;
                if (d < r + EPS)
                    list.add(new data(d, i, j));
            }
        }
        list.add(new data(r, 0, 0));
        Collections.sort(list);
        int cnt = 0;
        for (data d : list) {
            for (int j = 1; j <= m; j++)
                skip[match[j]] = true;
            for (int j = 1; j <= n; j++)
                if (!skip[j]) {
                    Arrays.fill(vis, false);
                    if (h(j))
                        cnt++;
                }
            int res = n + m - cnt;
            // printf("Radius %lf: count %d, ans %lf\n", dt[i].first, res, S(dt[i].first) *
            // res);
            ans = Math.max(ans, S(d.r) * res);
            ban[d.x][d.y] = true;
        }
        return ans;
    }
}